### PR TITLE
Fix AABB tree performance.

### PR DESCRIPTION
**Pull request by joaander**
## Description

Use the fast AABB-AABB overlap check when traversing AABB trees.

I implemented this by moving the `overlap(AABB, AABB)` function to a member function of AABB: `aabb.overlaps(other_aabb)`. Explicitly calling the intended overlap test ensures that the correct function is called.

## Motivation and context

Increase performance.

For reasons I don't fully understand, the compiler chose to automatically convert AABBs to OBBs, then call `overlap(OBB, OBB)` - which is many times slower than `overlap(AABB, AABB)`.

## How has this been tested?

This pull request increases performance by ~40-50% in HPMC benchmarks (https://github.com/glotzerlab/hoomd-benchmarks/pull/68). It should also improve MD performance with `nlist.Tree`, though I haven't tested that.

## Change log

```
Fixed:

* Increase HPMC performance a further ~50% on the CPU.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: hoomd/hpmc/OBB.h [hunk 1]
```diff
@@ -115,7 +115,7 @@ struct OBB
         is_sphere = 1;
         }
 
-    DEVICE OBB(const hoomd::detail::AABB& aabb)
+    DEVICE explicit OBB(const hoomd::detail::AABB& aabb)
```
Comment (joaander) [inline hoomd/hpmc/OBB.h:118]: This change ensures that no future automatic conversions from AABB to OBB are allowed.

Callers must opt into the conversion explicitly: `OBB obb(aabb)` or `OBB(aabb)`.

**Review by tommy-waltmann — COMMENTED** [review]
Have you tested to see if it was necessary to move the function call to a class member of AABB, or if just adding `explicit` to the OBB constructor was enough to do the trick?

**Review by tommy-waltmann — APPROVED** [review]
Approving because I don't want to hold up the next release.

**Review by tcmoore3 — APPROVED** [review]

**Comment by joaander** [comment]
> Have you tested to see if it was necessary to move the function call to a class member of AABB, or if just adding `explicit` to the OBB constructor was enough to do the trick?

When I only added explicit, I got an error compile error that there was no overlap(AABB, AABB) to call. I don't know why because it is clearly in the header file - so I moved it to a member instead.

**Comment by tommy-waltmann** [comment]
> When I only added explicit, I got an error compile error that there was no overlap(AABB, AABB) to call. I don't know why because it is clearly in the header file - so I moved it to a member instead.

That seems like the reason the compiler was automatically converting the AABBs to OBBs, it couldn't find the overlap(AABB, AABB) declaration.

**Pull request merged**